### PR TITLE
Add OnBeforeSave for metadata. Touch #5

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/JimBobSquarePants/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practise as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
This PR adds a new configurable action to the processing pipeline `OnBeforeSave` which passes the current `FormattedImage` instance for final adjustment before encoding to the output stream. This allows manipulation of EXIF metadata or anything else you can think of that cannot be handled via the URL API.

### CODE REVIEW

File: src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs [hunk 1]
```diff
@@ -174,6 +174,7 @@ public class ImageSharpMiddleware
                 using (var image = FormattedImage.Load(this.options.Configuration, inBuffer))
                 {
                     image.Process(this.logger, this.processors, commands);
+                    this.options.OnBeforeSave(image);
```
Comment (tocsoft) [inline src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs]: probably want to call this using this pattern `this.options.OnBeforeSave?.Invoke(image);` in the event `OnBeforeSave` is ever manually set to null.

Comment (tocsoft) [reply src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs]: We probably want to make sure we do that for all the extension points form the `options`

Comment (JimBobSquarePants) [reply src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs]: I don't like the idea of swallowing up exceptions like that. If someone manually sets it to null it should throw. 

I'm actually following the pattern from the MS staticfile middleware.

Comment (tocsoft) [reply src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs]: Its not an exception really as you wouldn't be swallowing anything its just skipping running the delegate if its but unset.

Comment (JimBobSquarePants) [reply src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs]: Here's what I mean. 

https://github.com/aspnet/StaticFiles/blob/826e89a2d1b325485819040ff41b9948afc91e06/src/Microsoft.AspNetCore.StaticFiles/StaticFileContext.cs#L285

Can't find a unit test to copy either. We do need to do some others though.

**Comment by JimBobSquarePants** [comment]
@tocsoft Finally got some decent integration tests running. Fixes some bugs in the process.

**Pull request merged**